File: benchmark/bench.dart
import 'dart:convert';
import 'dart:typed_data';

import '../lib/vdf.dart';

const List<int> _proveDifficulties = <int>[500, 1000, 10000, 1000000];
const List<int> _verifyDifficulties = <int>[500, 1000, 10000];

const Map<int, int> _defaultProveIterations = <int, int>{
  500: 8,
  1000: 6,
  10000: 2,
  1000000: 1,
};

const Map<int, int> _defaultVerifyIterations = <int, int>{
  500: 200,
  1000: 150,
  10000: 60,
};

void main(List<String> args) {
  final multiplier = args.isNotEmpty ? int.parse(args.first) : 1;
  if (multiplier <= 0) {
    throw ArgumentError('iteration multiplier must be positive');
  }

  final payload = Uint8List.fromList(utf8.encode('benchmark-payload'));

  print('Dart Wesolowski VDF benchmark');
  print('iteration multiplier: $multiplier');
  print('');

  _benchmarkProve(payload, multiplier);
  print('');
  _benchmarkVerify(payload, multiplier);
}

void _benchmarkProve(Uint8List payload, int multiplier) {
  print('BenchmarkProve');

  for (final difficulty in _proveDifficulties) {
    final iterations = (_defaultProveIterations[difficulty] ?? 1) * multiplier;
    final vdf = _benchmarkVdf();

    final sw = Stopwatch()..start();
    for (var i = 0; i < iterations; i++) {
      vdf.prove(payload, difficulty);
    }
    sw.stop();

    final totalMs = sw.elapsedMicroseconds / 1000.0;
    final avgMs = totalMs / iterations;

    print(
      '  difficulty=$difficulty iterations=$iterations '
      'total_ms=${totalMs.toStringAsFixed(3)} avg_ms=${avgMs.toStringAsFixed(3)}',
    );
  }
}

void _benchmarkVerify(Uint8List payload, int multiplier) {
  print('BenchmarkVerify');

  for (final difficulty in _verifyDifficulties) {
    final iterations = (_defaultVerifyIterations[difficulty] ?? 1) * multiplier;

    final prover = _benchmarkVdf();
    final proof = prover.prove(payload, difficulty);
    final verifier = Wesolowski.withPublicParams(prover.publicParams());

    final sw = Stopwatch()..start();
    for (var i = 0; i < iterations; i++) {
      final ok = verifier.verify(payload, difficulty, proof);
      if (!ok) {
        throw StateError('verification failed at difficulty=$difficulty');
      }
    }
    sw.stop();

    final totalMs = sw.elapsedMicroseconds / 1000.0;
    final avgMs = totalMs / iterations;

    print(
      '  difficulty=$difficulty iterations=$iterations '
      'total_ms=${totalMs.toStringAsFixed(3)} avg_ms=${avgMs.toStringAsFixed(3)}',
    );
  }
}

Wesolowski _benchmarkVdf() {
  return Wesolowski.withModulus(_benchmarkModulus(), 128);
}

BigInt _benchmarkModulus() {
  final p = (BigInt.one << 521) - one;
  final q = (BigInt.one << 607) - one;
  return p * q;
}

File: src/bin/bench.rs
use optimized_vdf::{Proof, Wesolowski};
use std::time::Instant;

fn main() {
    let difficulties = parse_difficulties();
    let payload = b"benchmark-payload";
    let vdf = benchmark_vdf();

    println!("Rust Wesolowski VDF benchmark");
    println!("BenchmarkProve");
    let mut proofs = Vec::with_capacity(difficulties.len());
    for difficulty in &difficulties {
        let start = Instant::now();
        let proof = vdf
            .prove(payload, *difficulty)
            .unwrap_or_else(|err| panic!("prove difficulty={difficulty}: {err:?}"));
        let elapsed = start.elapsed();
        println!(
            "  difficulty={} iterations=1 total_ms={:.3} avg_ms={:.3}",
            difficulty,
            elapsed.as_secs_f64() * 1000.0,
            elapsed.as_secs_f64() * 1000.0
        );

        verify_once(&vdf, payload, *difficulty, &proof);
        proofs.push(proof);
    }

    println!();
    println!("BenchmarkVerify");
    for (difficulty, proof) in difficulties.iter().zip(proofs.iter()) {
        let start = Instant::now();
        verify_once(&vdf, payload, *difficulty, proof);
        let elapsed = start.elapsed();
        println!(
            "  difficulty={} iterations=1 total_ms={:.3} avg_ms={:.3}",
            difficulty,
            elapsed.as_secs_f64() * 1000.0,
            elapsed.as_secs_f64() * 1000.0
        );
    }
}

fn parse_difficulties() -> Vec<i64> {
    let args: Vec<String> = std::env::args().skip(1).collect();
    if args.is_empty() {
        return vec![500, 1000, 10_000, 1_000_000];
    }

    args.iter()
        .map(|arg| {
            let value = arg
                .parse::<i64>()
                .unwrap_or_else(|err| panic!("invalid difficulty {arg:?}: {err}"));
            if value < 0 {
                panic!("difficulty must be non-negative, got {value}");
            }
            value
        })
        .collect()
}

fn benchmark_vdf() -> Wesolowski {
    Wesolowski::with_modulus(benchmark_modulus(), 128).expect("benchmark modulus")
}

fn benchmark_modulus() -> num_bigint::BigUint {
    let p = (num_bigint::BigUint::from(1u8) << 521usize) - num_bigint::BigUint::from(1u8);
    let q = (num_bigint::BigUint::from(1u8) << 607usize) - num_bigint::BigUint::from(1u8);
    p * q
}

fn verify_once(vdf: &Wesolowski, payload: &[u8], difficulty: i64, proof: &Proof) {
    let ok = vdf
        .verify(payload, difficulty, proof)
        .unwrap_or_else(|err| panic!("verify difficulty={difficulty}: {err:?}"));
    if !ok {
        panic!("verification failed at difficulty={difficulty}");
    }
}
